go/pkg/reader/reader.go

```diff
@@ -187,55 +193,34 @@ func NewCompressedSeeker(fs ReadSeeker) (ReadSeeker, error) {
 		fs:    fs,
 		encdW: encdW,
 		buf:   sb,
-		err:   nil,
 	}, nil
 }
 
 func (cfs *compressedSeeker) Read(p []byte) (int, error) {
-	if cfs.err != nil {
-		return 0, cfs.err
-	}
-
-	var n int
-	var errR, errW error
-	for cfs.encdW != nil && cfs.buf.Len() < len(p) && errR == nil && errW == nil {
-		// Read is allowed to use the entity of p as a scratchpad.
-		n, errR = cfs.fs.Read(p)
-		// errW must be non-nil if written bytes != from n.
-		_, errW = cfs.encdW.Encoder.Write(p[:n])
-	}
-
-	// We have to treat EOF differently. It's the only "everything is going according
-	// to plan" error. EOF implies our uncompressed input data has finished reading,
-	// but it doesn't mean that we've finished preparing/returning our compressed data.
-	// We only want to return EOF to the user when the _compressed_ data is finished reading.
-	if errR != nil && errR != io.EOF {
-		cfs.err = errR
-	} else if errW != nil {
-		cfs.err = errW
-	}
-
-	// When the buffer ends, or in case of _any_ error, we compress all the bytes we
-	// had available. The encoder requires a Close call to finish writing compressed
-	// data smaller than zstd's window size.
-	var errC error
-	if cfs.err != nil || errR == io.EOF {
-		errC = cfs.encdW.Encoder.Close()
+	if !cfs.IsInitialized() {
+		return 0, errNotInitialized
+	}
+
+	var err error
+	if cfs.encdW != nil && cfs.buf.Len() < len(p) {
+		// Read is allowed to use the entirety of p as a scratchpad.
+		_, err = io.CopyBuffer(cfs.encdW.Encoder, cfs.fs, p)
```

Comment:
It seems that this line would copy the entire content of the reader into the underlying buffer of the encoder (the provided buffer is used for staging, but does not limit the number of bytes copied). However, only up to `len(p)` is actually returned to the caller. A subsequent call to the method will simply continue to fill the specified buffer from the cached content of the first call. This could lead to OOM errors. [`io.CopyN`](https://pkg.go.dev/io#CopyN) might be a more suitable API here.

Reply:
`io.CopyN` doesn't swallow `io.EOF` the same way that `io.Copy` and `io.CopyBuffer` do. It also doesn't take a buffer to use as scratch like `io.CopyBuffer` does. That said, the implementation is pretty simple.

Could we try to adapt this to use `io.CopyBuffer` + `io.LimitReader`?

```
_, err = io.CopyBuffer(cfs.encdW.Encoder, io.LimitReader(cfs.fs, len(p)), p)
```

Reply:
I think that expression was wrong. It should be
```
_, err = io.CopyBuffer(cfs.encdW.Encoder, io.LimitReader(cfs.fs, len(p) - cfs.buf.Len()), p)
```

Reply:
I still can't get `CopyBuffer` working even with `LimitReader`. I could continue digging into it, but at this point I'd prefer to submit the PR as-is, especially because I'd rather limit the scope of the refactoring that's bundled with the bug fix.